cmd/tsbs_generate_data/serialize/akumuli.go

```
@@ -0,0 +1,136 @@
+	// Add cue
+	const HeaderLength = 8
+	buf = append(buf, "AAAAFFEE+"...)
+
+	// Series name
+	for i := 0; i < len(p.fieldKeys); i++ {
+		buf = append(buf, p.measurementName...)
+		buf = append(buf, '.')
+		buf = append(buf, p.fieldKeys[i]...)
+		if i+1 < len(p.fieldKeys) {
+			buf = append(buf, '|')
+		} else {
+			buf = append(buf, ' ')
+		}
+	}
+
+	for i := 0; i < len(p.tagKeys); i++ {
+		buf = append(buf, ' ')
+		buf = append(buf, p.tagKeys[i]...)
+		buf = append(buf, '=')
+		buf = append(buf, p.tagValues[i].(string)...)
+	}
+
+	series := string(buf[HeaderLength:])
+	if !s.bookClosed {
+		// Save point for later
+		if id, ok := s.book[series]; ok {
+			s.bookClosed = true
+			_, err = w.Write(s.deferred)
+			if err != nil {
+				return err
+			}
+			buf = buf[:HeaderLength]
+			buf = append(buf, fmt.Sprintf(":%d", id)...)
+			binary.LittleEndian.PutUint32(buf[:4], id)
+		} else {
+			// Shortcut
+			s.index++
+			tmp := make([]byte, 0, 1024)
+			tmp = append(tmp, "AAAAFFEE*2\n"...)
+			tmp = append(tmp, buf[HeaderLength:]...)
+			tmp = append(tmp, '\n')
+			tmp = append(tmp, fmt.Sprintf(":%d\n", s.index)...)
+			s.book[series] = s.index
+			// Update cue
+			binary.LittleEndian.PutUint16(tmp[4:6], uint16(len(tmp)))
+			binary.LittleEndian.PutUint16(tmp[6:HeaderLength], uint16(0))
+			binary.LittleEndian.PutUint32(tmp[:4], s.index)
+			binary.LittleEndian.PutUint32(buf[:4], s.index)
+			_, err = w.Write(tmp)
+			if err != nil {
+				return err
+			}
+			deferPoint = true
+			buf = buf[:HeaderLength]
+			buf = append(buf, fmt.Sprintf(":%d", s.index)...)
+		}
+	} else {
+		// Replace the series name with the value from the book
+		if id, ok := s.book[series]; ok {
+			buf = buf[:HeaderLength]
+			buf = append(buf, fmt.Sprintf(":%d", id)...)
+			binary.LittleEndian.PutUint16(buf[4:6], uint16(len(buf)))
+			binary.LittleEndian.PutUint16(buf[6:HeaderLength], uint16(0))
+			binary.LittleEndian.PutUint32(buf[:4], id)
+		} else {
+			panic("Unexpected series name")
```

let's just return an error here? A long term goal I have is to reduce the amount of panics in the codebase, and given this func already returns errors, we should probably just do that

---

Done